src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+open HolKernel Parse boolLib bossLib;
+
+(* From /core: *)
+open bir_programTheory bir_typing_progTheory bir_envTheory
+     bir_auxiliaryTheory bir_valuesTheory bir_expTheory
+     bir_exp_immTheory bir_typing_expTheory bir_immTheory;
+
+(* From /theories: *)
+open bir_program_blocksTheory bir_program_terminationTheory
+     bir_program_valid_stateTheory bir_exp_substitutionsTheory
+     bir_bool_expTheory bir_program_env_orderTheory
+     bir_program_multistep_propsTheory bir_exp_equivTheory
+     bir_subprogramTheory;
+
+open bir_wpTheory;
+
+open HolBACoreSimps;
+
+val _ = new_theory "bir_ht";
+
+val bir_stmtb_is_not_assume_def = Define `
+  (bir_stmtb_is_not_assume (BStmt_Declare v) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assign v ex) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assert ex) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assume ex) = F) /\
+  (bir_stmtb_is_not_assume (BStmt_Observe ec el obf) = T)
+`;
+
+val bir_stmtb_not_assume_never_assviol =
+  store_thm("bir_stmtb_not_assume_never_assviol",
+  ``!stmtb st obs st'.
+    bir_stmtb_is_not_assume stmtb ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_stmtB stmtb st = (obs, st')) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+REPEAT STRIP_TAC >>
+Cases_on `stmtb` >> (
+  FULL_SIMP_TAC std_ss [bir_stmtb_is_not_assume_def,
+                        bir_exec_stmtB_def]
+) >| [
+  FULL_SIMP_TAC std_ss [bir_exec_stmt_declare_def] >>
+  Cases_on `bir_env_varname_is_bound st.bst_environ
+              (bir_var_name b)` >| [
+    FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+    FULL_SIMP_TAC std_ss [] >>
+    Cases_on `bir_env_update (bir_var_name b)
+                (bir_declare_initial_value (bir_var_type b))
+                (bir_var_type b)
+                st.bst_environ` >| [
+      FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+      RW_TAC std_ss [] >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [],
```

Comment:
You are using this simpset `(std_ss++holBACore_ss)` everywhere, shouldn't you create in once and use it instead of creating it for each use? This should also speed up the execution a bit.

Reply:
i'm not thinking about execution speed but it may be shorter and also become more readable. i will also try to apply this for future theories i think.

Reply:
I think that the readability gain depends on the name, because in this case the code is small so it's useful to know that the proof is using `std_ss` with `holBACore_ss`, instead of (e.g.) `pure_ss`.

PS: shoudn't `holBACore_ss` be named `HOLBA_CORE_ss` (upper-case name)?

Reply:
`(std_ss++holBACore_ss)` is super common in general. I suggest we abbreviate this to `std_holba_ss` directly in `HolBACoreSimps`.
 
> PS: shoudn't `holBACore_ss` be named `HOLBA_CORE_ss` (upper-case name)?

No idea, but the general pattern in HOL4 identifiers is that older stuff is in all caps, while newer stuff is all lowercase. HOL4 simpsets are all over the place.

What does @andreaslindner think?

Reply:
> `(std_ss++holBACore_ss)` is super common in general. I suggest we abbreviate this to `std_holba_ss` directly in `HolBACoreSimps`.

Should be `std_holba_core_ss`, but I'm not sure if that's necessary. It's quite easy to create the ones we want in the top or our proofs. For example, in my current proof I included some `wordsTheory` and `pred_setTheory` ssfrags in my "`holba_ss`" simpset.


> > PS: shoudn't `holBACore_ss` be named `HOLBA_CORE_ss` (upper-case name)?

I was referring to [this](https://github.com/kth-step/HolBA/pull/50#pullrequestreview-237074280), I should have added the link. The idea is that by convention `ssfrag`s have upper-case names and `simpset` lower-case ones.

Reply:
> Should be `std_holba_core_ss`, but I'm not sure if that's necessary. It's quite easy to create the ones we want in the top or our proofs. For example, in my current proof I included some `wordsTheory` and `pred_setTheory` ssfrags in my "`holba_ss`" simpset.

True, but then we also have to contend with lots of identical abbreviations. The conclusion of this idea would be to redefine `std_holba_core_ss` in many files, since this combination is heavily used in a lot of places, now as well as going forward.

If we choose to go for a convenient shorthand, I also suggest we can even shorten it more, to `shc_ss` or something.

> I was referring to [this](https://github.com/kth-step/HolBA/pull/50#pullrequestreview-237074280), I should have added the link. The idea is that by convention `ssfrag`s have upper-case names and `simpset` lower-case ones.

Ah, I see. Thanks for the link.

Then I definitely agree we should rename `holBACore_ss`.

Reply:
> True, but then we also have to contend with lots of identical abbreviations. The conclusion of this idea would be to redefine `std_holba_core_ss` in many files, since this combination is heavily used in a lot of places, now as well as going forward.

Ok I agree, if that's really that used then we should define it  in HolBACoreSimps as you said. Then I wonder if we should also create a simpset with pred_set and words inside, because when we are simplifying expressions we expect (at least I do) that `(1w: word1 <> 0w: word1)` is simplified to `T`.

Reply:
>Then I wonder if we should also create a simpset with pred_set and words inside, because when we are simplifying expressions we expect (at least I do) that `(1w: word1 <> 0w: word1)` is simplified to `T`.

If it's just isolated small stuff from wordsTheory, then maybe you can make use of the lemmata in `bir_auxiliaryTheory` (`word1_distinct` and `word1_neg`). But I feel this is a performance question, I have no idea how brute-force it is in practice to use the simpsets derived from wordsTheory over individual theorems.

I think this discussion is bigger than this PR, can you create an issue about future simpsets?

Reply:
You're right, I created #57 and I'm resolving this.